README.md
# docker-samba
## About

[Samba](https://wiki.samba.org) Docker image based on Alpine Linux and [crazy-max docker-samba](https://github.com/crazy-max/docker-samba) repository.<br />
This container includes a web service discovery (wsdd, using `smbd`) to make the server discoverable by all Windows PCs on the same workgroup. Also, can recover custom user settings and passwords from samba TDB files (**T**rivial **D**ata**B**ase).

Instead build from scratch this container, you can pull it from [docker hub](https://hub.docker.com/r/begna112/docker-samba) or download only the `docker-compose.yml` and the `config.yml` file on [examples/compose/](examples/compose/data/config.yml).

___

- [docker-samba](#docker-samba)
  - [About](#about)
  - [Features](#features)
  - [Build locally](#build-locally)
  - [Environment variables](#environment-variables)
  - [Volumes](#volumes)
  - [Ports](#ports)
  - [Configuration](#configuration)
  - [Import custom user passwords and preferences](#import-custom-user-passwords-and-preferences)
  - [Usage](#usage)
    - [Docker Compose](#docker-compose)
    - [Command line](#command-line)
  - [Upgrade](#upgrade)
  - [Notes](#notes)
    - [Status](#status)
  - [Contributing](#contributing)
  - [License](#license)

## Features

* Easy [configuration](#configuration) through YAML
* Improve [operability with Mac OS X clients](https://wiki.samba.org/index.php/Configure_Samba_to_Work_Better_with_Mac_OS_X)
* This version mantains some support for legacy protocols including NetBIOS
* Backup and restore custom user passwords and settings

## Build locally

```shell
git clone https://github.com/Axia-SA/docker-samba.git
cd docker-samba

# Build image and output to docker (default)
docker buildx bake

# Build multi-platform image
docker buildx bake image-all
```

## Environment variables

* `TZ`: Timezone assigned to the container (default `UTC`)
* `SAMBA_WORKGROUP`: NT-Domain-Name or [Workgroup-Name](https://www.samba.org/samba/docs/current/man-html/smb.conf.5.html#WORKGROUP). (default `WORKGROUP`)
* `SAMBA_SERVER_STRING`: [Server string](https://www.samba.org/samba/docs/current/man-html/smb.conf.5.html#SERVERSTRING) is the equivalent of the NT Description field. (default `Docker Samba Server`)
* `SAMBA_LOG_LEVEL`: [Log level](https://www.samba.org/samba/docs/current/man-html/smb.conf.5.html#LOGLEVEL). (default `0`)
* `SAMBA_FOLLOW_SYMLINKS`: Allow to [follow symlinks](https://www.samba.org/samba/docs/current/man-html/smb.conf.5.html#FOLLOWSYMLINKS). (default `yes`)
* `SAMBA_WIDE_LINKS`: Controls whether or not links in the UNIX file system may be followed by the server. (default `yes`)
* `SAMBA_HOSTS_ALLOW`: Set of hosts which are permitted to access a service. (default `127.0.0.0/8 10.0.0.0/8 172.16.0.0/12 192.168.0.0/16`)
* `SAMBA_INTERFACES`: Allows you to override the default network interfaces list.
* `SAMBA_DISABLE_NETBIOS`: Disable NetBIOS for older connections. Values can be `yes` or `no` (default is `yes`)
* `SAMBA_CREATE_MASK`: Allows override of the default file permissions mask (default 0664)
* `SAMBA_FORCE_CREATE_MODE`: Forces override of the default file permissions mask (default 0664)
* `SAMBA_DIRECTORY_MASK`: Allows override of the default directory permissions mask (default 0775)
* `SAMBA_FORCE_DIRECTORY_MODE`: Forces override of the default directory permissions mask (default 0775)
> More info: https://www.samba.org/samba/docs/current/man-html/smb.conf.5.html

## Volumes

* `/data`: Contains cache, configuration and runtime data

## Ports

* `137/udp`: NetBIOS Name Service
* `138/udp`: NetBIOS Datagram
* `139/tcp`: NetBIOS Session
* `445/tcp`: Samba Server over TCP

> More info: https://wiki.samba.org/index.php/Samba_NT4_PDC_Port_Usage

## Configuration

Before using this image you have to create the YAML configuration file `/data/config.yml` to be able to create users,
provide global options and add shares. Here is an example:

```yaml
auth:
  - user: foo
    uid: 1000
    gid: 1000
    password: bar
  - user: baz
    uid: 1100
    gid: 1200
    password_file: /run/secrets/baz_password

global:
  - "force user = foo"
  - "force group = foo"

share:
  - name: foo
    path: /samba/foo
    browsable: yes
    readonly: no
    guestok: no
    validusers: foo
    writelist: foo
    veto: no
```

`veto: no` is a list of predefined files and directories that will not be
visible or accessible:

```
/._*/.apdisk/.AppleDouble/.DS_Store/.TemporaryItems/.Trashes/desktop.ini/ehthumbs.db/Network Trash Folder/Temporary Items/Thumbs.db/
```

More info: https://www.samba.org/samba/docs/current/man-html/smb.conf.5.html#VETOFILES

A more complete example is available [here](examples/compose/data/config.yml).

## Import custom user passwords and preferences

You can also use an existing TDB files to import previous custom settings like encrypted passwords that can not be stored on files under the `password_file` parameter on `config.yml`.

To achieve this, put the `passdb.tdb` and `secrets.tdb` files (or any `*.tdb` files) on `config/`. This files will be imported while the container is being built.

> Notice: TDB files must be consistent with `config.yml` file to work properly

## Usage

### Docker Compose

Docker compose is the recommended way to run this image. Copy the content of folder [examples/compose](examples/compose)
in `/var/samba/` on your host for example. Edit the compose and configuration files with your preferences and run the
following commands:

```bash
docker-compose up -d
docker-compose logs -f
```

### Command line

You can also use the following minimal command:

```shell
docker run -d --network host \
  -v "$(pwd)/data:/data" \
  --name samba cbottazzi/docker-samba
```

## Upgrade

Recreate the container when exists an update:

```bash
docker-compose pull
docker-compose up -d
```

## Notes
### Status

Use the following commands to check the logs and status:

```shell
docker-compose logs samba
docker-compose exec samba smbstatus
```

## Contributing

Want to contribute? Awesome! The most basic way to show your support is to star the project, or to raise issues.

Thanks again for your support, it is much appreciated! :pray:

## License

MIT. See `LICENSE` for more details.


CHANGELOG.md
# Changelog

## 4.16.7-r0 (2022/12/14)

* Samba 4.16.7
* Alpine Linux 3.17

## 4.15.2-r1 (2020/12/18)

* Fix `password_file` when password is unspecified (#10)

## 4.15.2-r0 (2020/12/17)

* Samba 4.15.2 (#9)
* Alpine Linux 3.15 (#9)
* Add `password_file` field (#8)

## 4.14.8-r0 (2020/10/27)

* Samba 4.14.8 (#7)

## 4.14.5-r1 (2020/07/14)

* Fix yq syntax

## 4.14.5-r0 (2020/07/14)

* Samba 4.14.5
* Alpine Linux 3.14

## 4.13.8-r1 (2020/05/30)

* Allow providing global options (#2)

## 4.13.8-r0 (2020/05/30)

* Initial version based on Samba 4.13.8


entrypoint.sh
#!/bin/bash

TZ=${TZ:-UTC}

SAMBA_WORKGROUP=${SAMBA_WORKGROUP:-WORKGROUP}
SAMBA_SERVER_STRING=${SAMBA_SERVER_STRING:-Docker Samba Server}
SAMBA_LOG_LEVEL=${SAMBA_LOG_LEVEL:-0}
SAMBA_FOLLOW_SYMLINKS=${SAMBA_FOLLOW_SYMLINKS:-yes}
SAMBA_WIDE_LINKS=${SAMBA_WIDE_LINKS:-yes}
SAMBA_HOSTS_ALLOW=${SAMBA_HOSTS_ALLOW:-127.0.0.0/8 10.0.0.0/8 172.16.0.0/12 192.168.0.0/16}
SAMBA_DISABLE_NETBIOS=${SAMBA_DISABLE_NETBIOS:-yes}
SAMBA_NETBIOS_NAME=${SAMBA_NETBIOS_NAME:-SMBSERVER}
#SAMBA_INTERFACES=${SAMBA_INTERFACES:-eth0}
SAMBA_CREATE_MASK=${SAMBA_CREATE_MASK:-0664}
SAMBA_FORCE_CREATE_MODE=${SAMBA_FORCE_CREATE_MODE:-0664}
SAMBA_DIRECTORY_MASK=${SAMBA_DIRECTORY_MASK:-0775}
SAMBA_FORCE_DIRECTORY_MODE=${SAMBA_DIRECTORY_MASK:-0775}

echo "Setting timezone to ${TZ}"
ln -snf /usr/share/zoneinfo/${TZ} /etc/localtime
echo ${TZ} > /etc/timezone

echo "Initializing files and folders"
mkdir -p /data/cache /data/lib
if [ -z "$(ls -A /data/lib)" ]; then
  cp -r /var/lib/samba/* /data/lib/
fi
rm -rf /var/lib/cache /var/lib/samba
ln -sf /data/cache /var/cache/samba
ln -sf /data/lib /var/lib/samba

echo "Setting global configuration"
  cat > /etc/samba/smb.conf <<EOL
[global]
workgroup = ${SAMBA_WORKGROUP}
server string = ${SAMBA_SERVER_STRING}
server role = standalone server
server services = -dns, -nbt
server signing = default
server multi channel support = yes

log level = ${SAMBA_LOG_LEVEL}
;log file = /usr/local/samba/var/log.%m
;max log size = 50

hosts allow = ${SAMBA_HOSTS_ALLOW}
hosts deny = 0.0.0.0/0

security = user
guest account = nobody
pam password change = yes
map to guest = bad user
usershare allow guests = yes

create mask = ${SAMBA_CREATE_MASK)
force create mode = ${SAMBA_FORCE_CREATE_MODE}
directory mask = ${SAMBA_DIRECTORY_MASK}
force directory mode = ${SAMBA_DIRECTORY_MASK}
follow symlinks = ${SAMBA_FOLLOW_SYMLINKS}
wide links = ${SAMBA_WIDE_LINKS}
unix extensions = no

printing = bsd
printcap name = /dev/null
disable spoolss = no
disable netbios = ${SAMBA_DISABLE_NETBIOS}
netbios name = ${SAMBA_NETBIOS_NAME}
smb ports = 445

client ipc min protocol = default
client ipc max protocol = default

ntlm auth = yes
wins support = yes
local master = yes
preferred master = yes

;wins server = w.x.y.z
;wins proxy = yes
dns proxy = no
socket options = TCP_NODELAY
strict locking = no

winbind scan trusted domains = yes

EOL

if [ -n "${SAMBA_INTERFACES}" ]; then
  cat >> /etc/samba/smb.conf <<EOL
interfaces = ${SAMBA_INTERFACES}
bind interfaces only = yes

EOL
fi

if [[ "$(yq -j e /data/config.yml 2>/dev/null | jq '.auth')" != "null" ]]; then
  for auth in $(yq -j e /data/config.yml 2>/dev/null | jq -r '.auth[] | @base64'); do
    _jq() {
      echo "${auth}" | base64 --decode | jq -r "${1}"
    }
    password=$(_jq '.password')
    if [[ "$password" = "null" ]] && [[ -f "$(_jq '.password_file')" ]]; then
      password=$(cat "$(_jq '.password_file')")
    fi
    echo "Creating user $(_jq '.user')/$(_jq '.group') ($(_jq '.uid'):$(_jq '.gid'))"
    id -g "$(_jq '.gid')" &>/dev/null || id -gn "$(_jq '.group')" &>/dev/null || addgroup -g "$(_jq '.gid')" -S "$(_jq '.group')"
    id -u "$(_jq '.uid')" &>/dev/null || id -un "$(_jq '.user')" &>/dev/null || adduser -u "$(_jq '.uid')" -G "$(_jq '.group')" "$(_jq '.user')" -SHD
    echo -e "$password\n$password" | smbpasswd -a -s "$(_jq '.user')"
    unset password
  done
fi

if [[ "$(yq -j e /data/config.yml 2>/dev/null | jq '.global')" != "null" ]]; then
  for global in $(yq -j e /data/config.yml 2>/dev/null | jq -r '.global[] | @base64'); do
  echo "Add global option: $(echo "$global" | base64 --decode)"
  cat >> /etc/samba/smb.conf <<EOL
$(echo "$global" | base64 --decode)
EOL
  done
fi

if [[ "$(yq -j e /data/config.yml 2>/dev/null | jq '.share')" != "null" ]]; then
  for share in $(yq -j e /data/config.yml 2>/dev/null | jq -r '.share[] | @base64'); do
    _jq() {
      echo "${share}" | base64 --decode | jq -r "${1}"
    }
    echo "Creating share $(_jq '.name')"
    if [[ "$(_jq '.name')" = "null" ]] || [[ -z "$(_jq '.name')" ]]; then
      >&2 echo "ERROR: Name required"
      exit 1
    fi
    echo -e "\n[$(_jq '.name')]" >> /etc/samba/smb.conf
    if [[ "$(_jq '.path')" = "null" ]] || [[ -z "$(_jq '.path')" ]]; then
      >&2 echo "ERROR: Path required"
      exit 1
    fi
    echo "path = $(_jq '.path')" >> /etc/samba/smb.conf
    if [[ "$(_jq '.comment')" != "null" ]] && [[ -n "$(_jq '.comment')" ]]; then
      echo "comment = $(_jq '.comment')" >> /etc/samba/smb.conf
    fi
    if [[ "$(_jq '.browsable')" = "null" ]] || [[ -z "$(_jq '.browsable')" ]]; then
      echo "browsable = yes" >> /etc/samba/smb.conf
    else
      echo "browsable = $(_jq '.browsable')" >> /etc/samba/smb.conf
    fi
    if [[ "$(_jq '.readonly')" = "null" ]] || [[ -z "$(_jq '.readonly')" ]]; then
      echo "read only = yes" >> /etc/samba/smb.conf
    else
      echo "read only = $(_jq '.readonly')" >> /etc/samba/smb.conf
    fi
    if [[ "$(_jq '.guestok')" = "null" ]] || [[ -z "$(_jq '.guestok')" ]]; then
      echo "guest ok = yes" >> /etc/samba/smb.conf
    else
      echo "guest ok = $(_jq '.guestok')" >> /etc/samba/smb.conf
    fi
    if [[ "$(_jq '.validusers')" != "null" ]] && [[ -n "$(_jq '.validusers')" ]]; then
      echo "valid users = $(_jq '.validusers')" >> /etc/samba/smb.conf
    fi
    if [[ "$(_jq '.adminusers')" != "null" ]] && [[ -n "$(_jq '.adminusers')" ]]; then
      echo "admin users = $(_jq '.adminusers')" >> /etc/samba/smb.conf
    fi
    if [[ "$(_jq '.writelist')" != "null" ]] && [[ -n "$(_jq '.writelist')" ]]; then
      echo "write list = $(_jq '.writelist')" >> /etc/samba/smb.conf
    fi
    if [[ "$(_jq '.veto')" != "null" ]] && [[ "$(_jq '.veto')" = "no" ]]; then
      echo "veto files = /._*/.apdisk/.AppleDouble/.DS_Store/.TemporaryItems/.Trashes/desktop.ini/ehthumbs.db/Network Trash Folder/Temporary Items/Thumbs.db/" >> /etc/samba/smb.conf
      echo "delete veto files = yes" >> /etc/samba/smb.conf
    fi
  done
fi

testparm -s

echo "Executing nmbd to making network discovereable (wsdd - Web Service Discovery). Resource name: " ${SAMBA_NETBIOS_NAME}
nmbd --no-process-group --configfile /etc/samba/smb.conf

# -----------------------------------------------
echo "Recovery previous passdb.tdb database if exists"
FILE=/data/passdb.tdb
if test -f "$FILE"; then
    echo " -> $FILE exists. Starting recovery..."
    cp /data/*.tdb /data/lib/private/
    chmod -R 0700 /data/lib/private/
    echo "Done."
else 
    echo " -> $FILE does not exist. Skipping this step."
fi

# -----------------------------------------------

exec "$@"
